test/cpp/end2end/client_callback_end2end_test.cc

```diff
@@ -374,6 +374,58 @@ TEST_P(ClientCallbackEnd2endTest, SimpleRpc) {
   SendRpcs(1, false);
 }
 
+TEST_P(ClientCallbackEnd2endTest, SimpleRpcUnderLockNested) {
+  MAYBE_SKIP_TEST;
+  ResetStub();
+  std::mutex mu1, mu2, mu3;
+  std::condition_variable cv1, cv2, cv3;
+  bool done1 = false;
+  EchoRequest request1, request2, request3;
+  request1.set_message("Hello locked world1.");
+  request2.set_message("Hello locked world2.");
+  request3.set_message("Hello locked world3.");
+  EchoResponse response1, response2, response3;
+  ClientContext cli_ctx1, cli_ctx2, cli_ctx3;
+  {
+    std::unique_lock<std::mutex> l(mu1);
+    stub_->experimental_async()->Echo(
+        &cli_ctx1, &request1, &response1,
+        [this, &mu1, &mu2, &mu3, &cv1, &done1, &request1, &request2, &request3,
+         &response1, &response2, &response3, &cli_ctx1, &cli_ctx2,
+         &cli_ctx3](Status s1) {
+          std::unique_lock<std::mutex> l1(mu1);
+          EXPECT_TRUE(s1.ok());
+          EXPECT_EQ(request1.message(), response1.message());
+          // start the second level of nesting
+          std::unique_lock<std::mutex> l2(mu2);
+          this->stub_->experimental_async()->Echo(
+              &cli_ctx2, &request2, &response2,
+              [this, &mu2, &mu3, &cv1, &done1, &request2, &request3, &response2,
+               &response3, &cli_ctx3](Status s2) {
+                std::unique_lock<std::mutex> l2(mu2);
+                EXPECT_TRUE(s2.ok());
+                EXPECT_EQ(request2.message(), response2.message());
+                // start the third level of nesting
+                std::unique_lock<std::mutex> l3(mu3);
+                stub_->experimental_async()->Echo(
+                    &cli_ctx3, &request3, &response3,
+                    [&mu3, &cv1, &done1, &request3, &response3](Status s3) {
+                      std::lock_guard<std::mutex> l(mu3);
+                      EXPECT_TRUE(s3.ok());
+                      EXPECT_EQ(request3.message(), response3.message());
+                      done1 = true;
```

Comment:
There's a tsan race right now since you're editing this bool under mu3 but reading it under mu1 . Actually, I think you just need 1 mu and 1 cv, not 3 anymore.

Reply:
ohk. I kept the 3 locks around since I want to test the calling 3 gRPC callbacks under a lock and grabbing the same lock in the callback.. I have addressed the tsan issue by moving the boolean to under mu3.